src/main/scala/com/fulcrumgenomics/bam/LeftAlignIndels.scala

```diff
@@ -0,0 +1,330 @@
+      writer.write(rec)
+    }
+    writeProgress.logLast()
+
+    // Clean up
+    walker.foreach(_.close())
+    reader.safelyClose()
+    writer.close()
+
+    val pct = leftAligned * 100.0 / writeProgress.getCount
+    logger.info(f"Left aligned $leftAligned%,d out of ${writeProgress.getCount}%,d records ($pct%.2f%%)")
+  }
+}
+
+class LeftIndelAligner(useEqualsAndX: Boolean) {
+  /** Left-aligns the given [[SamRecord]]. */
+  def align(rec: SamRecord, walker: ReferenceSequenceFileWalker): Boolean = if (rec.unmapped) false else {
+    val refBases = walker.get(rec.refIndex).getBases.slice(from = rec.start-1, until = rec.end).map(_.toChar.toUpper).mkString("")
+    this.align(rec, refBases)
+  }
+
+  /** Left-aligns the given [[SamRecord]]. */
+  def align(rec: SamRecord, refBases: String): Boolean = if (rec.unmapped) false else {
+    val original    = IndelPaddedAlignment(rec, refBases)
+    val leftAligned = original.leftAligned
+    val cigar       = leftAligned.cigar(useEqualsAndX=useEqualsAndX)
+
+    // Set the cigar
+    rec.cigar = {
+      val leadingHardClip  = rec.cigar.takeWhile(_.operator == CigarOperator.H).map(_.length).sum
+      val trailingHardClip = rec.cigar.elems.drop(1).reverse.takeWhile(_.operator == CigarOperator.H).map(_.length).sum
+      (leadingHardClip, trailingHardClip) match {
+        case (0, 0) => cigar
+        case (n, 0) => new Cigar(CigarElem(CigarOperator.H, n) +: cigar.elems)
+        case (0, m) => new Cigar(cigar.elems :+ CigarElem(CigarOperator.H, m))
+        case (n, m) => new Cigar(CigarElem(CigarOperator.H, n) +: cigar.elems :+ CigarElem(CigarOperator.H, m))
+      }
+    }
+
+    original == leftAligned
+  }
+}
+
+
+object IndelPaddedAlignment {
+  /** The character in the read or reference for insertions or deletions respectively. */
+  val Indel: Char = '-'
+  /** The character in the reference to indicate the read was soft-clipped. */
+  val SoftClip: Char = 'X'
+
+  // Convenience values
+  private val IndelString: String = Indel.toString
+  private val SoftClipString: String = SoftClip.toString
+
+  /** Builds a padded alignment for the given [[SamRecord]] and its corresponding target bases.  The record must be
+    * mapped. NB: hard-clipped bases are lost. */
+  def apply(rec: SamRecord, target: String): IndelPaddedAlignment = {
+    require(rec.mapped)
+    this.apply(rec.cigar, rec.basesString, target)
+  }
+
+  /** Builds a padded alignment for the given [[Cigar]], query and target target bases. NB: hard-clipped bases are lost. */
+  def apply(cigar: Cigar, query: String, target: String): IndelPaddedAlignment = {
+    require(cigar.lengthOnQuery == query.length,
+      s"query length mismatch: ${cigar.lengthOnQuery} vs. ${query.length}")
+    require(cigar.lengthOnTarget == target.length,
+      s"target length mismatch: ${cigar.lengthOnTarget} vs. ${target.length}")
+
+    val queryBuilder  = new ArrayBuffer[Char]()
+    val targetBuilder = new ArrayBuffer[Char]()
```

Comment:
Not perfect, but better
```suggestion
    val queryBuilder  = new ArrayBuffer[Char](query.length)
    val targetBuilder = new ArrayBuffer[Char](target.length)
```